# 3D de novo assembly (3D-DNA) pipeline, ENCODE phasing branch

This version of the pipeline (211104) adds a DCC Juicer2 Hi-C processing pipeline-compatible phasing module.

### Overview of the ENCODE phasing module
`Software version: 211104`

The wrapper script for variant phasing using the DCC Juicer2  is located in the phasing folder of the 3D-DNA pipeline:
`bash 3d-dna/phasing/run-hic-phaser-encode.sh`

##### ******************************

#### USAGE:
`bash ./run-hic-phaser-encode.sh [options] <path_to_vcf> <path_to_aligned_dedupped_bam>`

#### DESCRIPTION:
This is a wrapper script to use 3D-DNA phaser to phase SNPs using Hi-C alignment data as generated by the ENCODE DCC hic & variant calling pipeline.

#### ARGUMENTS:
`path_to_vcf`
						Path to a Variant Call Format (vcf) file containing sequence variation data, e.g. as generated by the ENCODE DCC Hi-C variant calling pipeline.

`path_to_aligned_dedupped_bam`
						Path to bam file containing deduplicated alignments of Hi-C reads in bam format (output by Juicer2).

#### OPTIONS:
`-h|--help`
						Shows help.

##### DATA FILTERING:						
`-c|--chr [chr_list]`
		                Phase only specific molecules (default phases all molecules that have variants listed in the vcf file). Note that -c and -C are incompatible in that if both are invoked, only the last listed option will be taken into account.

`-C|--exclude-chr [chr_list]`
						Remove specific molecules from the chromosome list (default: chrY). Note that -c and -C are incompatible in that if both are invoked, only the last listed option will be taken into account.

`-q|--mapq [mapq]`
		                Consider only Hi-C reads that align with minimal mapping quality of mapq (default is 1).

##### PHASER CONTROL:
`-s|--stringency [stringency]`
		                Specify stringency parameter for the phaser (default is 3, i.e. 3-fold enrichment in Hi-C signal to one molecule as compared to the other is necessary to phase).

`-b|--background [background]`
		                Specify background parameter for the phaser (default is 1, i.e. calculate enrichment on top of the noise level of 1 read)

`-v|--verbose [step_for_intermediate_data_dump]`
						Print intermediate phasing results every [step_for_intermediate_data_dump] steps.

##### OUTPUT CONTROL:
`--separate-homolog-maps`
						Build two separate contact maps for homologs as opposed to a single diploid contact map with interleaved homologous chromosomes. This is the preferred mode for comparing contacts across homologs using the \"Observed over Control\" view for map comparison in Juicebox. Importantly, assignment of chromosomes to homologs is arbitrary. Default: not invoked. 

##### WORKFLOW CONTROL:
`-t|--threads [num]`
        				Indicate how many threads to use. Default: half of available cores as calculated by parallel --number-of-cores.

`--from-stage [pipeline_stage]`
						Fast-forward to a particular stage of the pipeline. The pipeline_stage argument can be \"prep\", \"parse_vcf\", \"parse_bam\", \"visualize_input\", \"phase\", \"visualize_output\", \"update_vcf\", \"map_diploid\", \"cleanup\".

`--to-stage [pipeline_stage]`
						Exit after a particular stage of the pipeline. The argument can be \"prep\", \"parse_vcf\", \"parse_bam\", \"visualize_input\", \"phase\", \"visualize_output\", \"update_vcf\", \"map_diploid\", \"cleanup\".

##### ******************************

Typical usage would invoke
- 1) Running Juicer2 on all samples associated with a particular individual sample and creating a 'mega' merged_dedupped.bam file.
- 2) Running [hic2gatk](https://github.com/aidenlab/hic2gatk) script to call SNPs from the alignments produced by the Juicer2 pipeline (snp.out.vcf file).
- 3) Running 3d-dna/phasing/run-hic-phaser-encode.sh snp.out.vcf merged_dedupped.bam to generate a snp.out_HiC.vcf file with phasing information added in the form of a PS tag. In addition to the vcf file several Hi-C contact maps are created including the diploid.hic. For human samples the latter map contains 46 intra-molecular submaps, two homologs per each of the 23 chromosomes (chr1-r, chr1-a; chr2-r, chr2-a etc.).

Note that phasing requires considerably higher coverages as compared to SNP calling, so it is recommended that, if available, all data for an individual is used for phasing, including that spanning different tissues. If needed, creating tissue-specific diploid.hic maps using the joint phasing information can be done by:

- 1) making an archive copy of the reads.sorted.bam file (and of associated index), one of the files generated by the prep stage of the phasing pipeline and containing sorted unique paired alignments.
- 2) running `samtools view -R FILE reads.sorted.archive.bam > reads.sorted.bam` where FILE lists the read groups relevant to the specific tissue.
- 3) running the map_diploid stage of the phasing pipeline to generate the phased maps containing data only from the experiments associated with the read groups in FILE.
